File: packages/shared/src/settlement.ts
import type { SettlementOutputs, SettlementRates, SettlementPhase } from "./types";
import { calculateRevenue } from "./energy";

const round = (value: number) => Number(value.toFixed(2));

/** Suggested recovery-phase split (20 parts) — see docs/imported-specs/installation-process-drs-lbrs-go-live.md §5.1 */
export const RECOVERY_WATERFALL_RATES: SettlementRates = {
  reserve: 1 / 20,
  providers: 6 / 20,
  financiers: 10 / 20,
  owner: 1 / 20,
  emappa: 2 / 20,
};

/** Suggested royalty-phase split (20 parts) */
export const ROYALTY_WATERFALL_RATES: SettlementRates = {
  reserve: 2 / 20,
  providers: 7 / 20,
  financiers: 2 / 20,
  owner: 2 / 20,
  emappa: 7 / 20,
};

export function defaultWaterfallRates(phase: SettlementPhase): SettlementRates {
  return phase === "recovery" ? { ...RECOVERY_WATERFALL_RATES } : { ...ROYALTY_WATERFALL_RATES };
}

export function calculateSettlement(
  E_sold: number,
  pricePerKwh: number,
  rates: SettlementRates,
  options?: { phase?: SettlementPhase },
): SettlementOutputs {
  const phase = options?.phase ?? "recovery";
  void phase; // reserved for future phase-specific rules beyond rates passed in
  const revenue = calculateRevenue(E_sold, pricePerKwh);
  const reserve = revenue * rates.reserve;
  const providerPool = revenue * rates.providers;
  const financierPool = revenue * rates.financiers;
  const ownerRoyalty = revenue * rates.owner;
  const emappaFee = revenue * rates.emappa;
  const allocated = reserve + providerPool + financierPool + ownerRoyalty + emappaFee;
  const shortfallKes = round(Math.max(0, allocated - revenue));
  const scale = allocated > revenue && allocated > 0 ? revenue / allocated : 1;
  const reserveA = round(reserve * scale);
  const providerA = round(providerPool * scale);
  const financierA = round(financierPool * scale);
  const ownerA = round(ownerRoyalty * scale);
  const emappaA = round(emappaFee * scale);
  const allocatedTotal = round(reserveA + providerA + financierA + ownerA + emappaA);
  const unallocated = round(revenue - allocatedTotal);

  return {
    revenue,
    reserve: reserveA,
    providerPool: providerA,
    financierPool: financierA,
    ownerRoyalty: ownerA,
    emappaFee: emappaA,
    unallocated,
    allocatedTotal,
    shortfallKes,
    phase: options?.phase ?? "recovery",
  };
}

export function validateSettlementRates(rates: SettlementRates) {
  const total = rates.reserve + rates.providers + rates.financiers + rates.owner + rates.emappa;
  return {
    total: round(total),
    isBalanced: Math.abs(total - 1) < 0.0001,
  };
}

File: backend/app/services/settlement.py
from .energy import calculate_revenue, round2


def calculate_settlement(e_sold: float, price_per_kwh: float, rates: dict, phase: str | None = None) -> dict:
    phase_resolved = phase if phase is not None else "recovery"
    revenue = calculate_revenue(e_sold, price_per_kwh)
    reserve = revenue * rates["reserve"]
    provider_pool = revenue * rates["providers"]
    financier_pool = revenue * rates["financiers"]
    owner_royalty = revenue * rates["owner"]
    emappa_fee = revenue * rates["emappa"]
    allocated = reserve + provider_pool + financier_pool + owner_royalty + emappa_fee
    scale = revenue / allocated if allocated > revenue and allocated > 0 else 1.0
    reserve_a = round2(reserve * scale)
    provider_a = round2(provider_pool * scale)
    financier_a = round2(financier_pool * scale)
    owner_a = round2(owner_royalty * scale)
    emappa_a = round2(emappa_fee * scale)
    allocated_total = round2(reserve_a + provider_a + financier_a + owner_a + emappa_a)
    unallocated = round2(revenue - allocated_total)
    shortfall = round2(max(0.0, allocated - revenue))
    return {
        "revenue": revenue,
        "reserve": reserve_a,
        "providerPool": provider_a,
        "financierPool": financier_a,
        "ownerRoyalty": owner_a,
        "emappaFee": emappa_a,
        "unallocated": unallocated,
        "allocatedTotal": allocated_total,
        "shortfallKes": shortfall,
        "phase": phase_resolved,
    }


def validate_settlement_rates(rates: dict) -> dict:
    total = rates["reserve"] + rates["providers"] + rates["financiers"] + rates["owner"] + rates["emappa"]
    return {"total": round2(total), "isBalanced": abs(total - 1) < 0.0001}
